barretenberg/cpp/src/barretenberg/vm/avm_trace/avm_kernel_trace.cpp

```diff
@@ -2,85 +2,246 @@
+}
+
+FF AvmKernelTraceBuilder::perform_kernel_input_lookup(uint32_t selector)
+{
+    FF result = std::get<0>(public_inputs)[selector];
+    kernel_input_selector_counter[selector]++;
     return result;
 }
 
+void AvmKernelTraceBuilder::perform_kernel_output_lookup(uint32_t write_offset, FF value, FF metadata)
+{
+    std::get<KERNEL_OUTPUTS_VALUE>(public_inputs)[write_offset] = value;
+    std::get<KERNEL_OUTPUTS_SIDE_EFFECT_COUNTER>(public_inputs)[write_offset] = side_effect_counter;
+    std::get<KERNEL_OUTPUTS_METADATA>(public_inputs)[write_offset] = metadata;
+
+    // Lookup counts
+    kernel_output_selector_counter[write_offset]++;
+
+    side_effect_counter++;
+}
+
 // We want to be able to get the return value from the public inputs column
 // Get the return value, this will be places in ia
 // We read from the public inputs that were provided to the kernel
 FF AvmKernelTraceBuilder::op_sender()
 {
-    return perform_kernel_lookup(SENDER_SELECTOR);
+    return perform_kernel_input_lookup(SENDER_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_address()
 {
-    return perform_kernel_lookup(ADDRESS_SELECTOR);
+    return perform_kernel_input_lookup(ADDRESS_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_portal()
 {
-    return perform_kernel_lookup(PORTAL_SELECTOR);
+    return perform_kernel_input_lookup(PORTAL_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_fee_per_da_gas()
 {
-    return perform_kernel_lookup(FEE_PER_DA_GAS_SELECTOR);
+    return perform_kernel_input_lookup(FEE_PER_DA_GAS_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_fee_per_l2_gas()
 {
-    return perform_kernel_lookup(FEE_PER_L2_GAS_SELECTOR);
+    return perform_kernel_input_lookup(FEE_PER_L2_GAS_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_transaction_fee()
 {
-    return perform_kernel_lookup(TRANSACTION_FEE_SELECTOR);
+    return perform_kernel_input_lookup(TRANSACTION_FEE_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_chain_id()
 {
-    return perform_kernel_lookup(CHAIN_ID_SELECTOR);
+    return perform_kernel_input_lookup(CHAIN_ID_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_version()
 {
-    return perform_kernel_lookup(VERSION_SELECTOR);
+    return perform_kernel_input_lookup(VERSION_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_block_number()
 {
-    return perform_kernel_lookup(BLOCK_NUMBER_SELECTOR);
+    return perform_kernel_input_lookup(BLOCK_NUMBER_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_coinbase()
 {
-    return perform_kernel_lookup(COINBASE_SELECTOR);
+    return perform_kernel_input_lookup(COINBASE_SELECTOR);
 }
 
 FF AvmKernelTraceBuilder::op_timestamp()
 {
-    return perform_kernel_lookup(TIMESTAMP_SELECTOR);
+    return perform_kernel_input_lookup(TIMESTAMP_SELECTOR);
+}
+
+// TODO(https://github.com/AztecProtocol/aztec-packages/issues/6481): need to process hint from avm in order to know if
+// output should be set to true or not
+void AvmKernelTraceBuilder::op_note_hash_exists(uint32_t clk, FF note_hash, uint32_t result)
```

Comment:
note_hash can be made const &
Same for the other methods: note_hash, nullifier, etc, ....